### PR TITLE
Assert if y_pred and y_true have same shape in categorical_accuracy()

### PR DESCRIPTION
**Pull request by AdityaKane2001**
Refer this [issue](https://github.com/tensorflow/tensorflow/issues/46953) for details.

### CODE REVIEW

**Comment by amogh7joshi** [comment]
@AdityaKane2001 Please check build failures, there seems to be a unittest that has `y_true` and `y_pred` such that their shapes are not equal.

**Comment by AdityaKane2001** [comment]
@amogh7joshi  Noted, but I don't know what to do with it. 😅 
If you see the aforementioned issue, you will see why one must assert that the shapes are equal. This is because the [`categorical_accuracy()` ](https://github.com/tensorflow/tensorflow/blob/85c8b2a817f95a3e979ecd1ed95bff1dc1335cff/tensorflow/python/keras/metrics.py#L3250)function just uses a plain `argmax` function. Thus, if the shapes are unequal, there will be no errors but the results may be wrong.

**Comment by amogh7joshi** [comment]
Agreed. The likelihood is that there is simply a test case which contains certain incompatible values, and you simply need to edit the test case, but there is also the (albeit unlikely) situation in which the test case *cannot* be updated at which point this feature will not work.

Regardless, you should check the build failures and update any necessary tests. Basically, click on "details" next to the failed test and scroll through the output log until you find the point where it says "FAILED" and something related to asserted shapes. Then, just update the test file which corresponds to it. Let me know if you have any other questions.

**Comment by AdityaKane2001** [comment]
@amogh7joshi 
I checked the logs earlier, and found that `test_saving_with_sequence_features_test_mode_graph (__main__.SequenceFeaturesSavingTest)` was failing. 
Where exactly can I find the mentioned files?

**Comment by AdityaKane2001** [comment]
@amogh7joshi 
I went through the logs. I found the module but I don't know what to update. Can you please guide a bit on the same? I am quite new to this...

File: tensorflow/python/keras/metrics.py [hunk 1]
```diff
@@ -3342,6 +3342,9 @@ def categorical_accuracy(y_true, y_pred):
   Returns:
     Categorical accuracy values.
   """
+  #assert if both have same shape
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/metrics.py:3345]: Nit: space between comment mark and text, and prefer a more specific comment:

\# assert if predicted and true labels have the same shape

File: tensorflow/python/keras/metrics.py [hunk 2]
```diff
@@ -3342,6 +3342,9 @@ def categorical_accuracy(y_true, y_pred):
...
+  assert y_true.shape==y_pred.shape,"Both inputs must have same shapes and must be one-hot encoded. Use tf.one_hot() to one hot encode the inputs."
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/metrics.py:3346]: y_true.shape returns the static shape, please replace with tf.shape(y_true), as tf.shape(..) resolves the actual shape at runtime.

File: tensorflow/python/keras/metrics.py (hunk 1 above)
Comment (AdityaKane2001) [reply tensorflow/python/keras/metrics.py:3345]: Noted.

File: tensorflow/python/keras/metrics.py (hunk 2 above)
Comment (AdityaKane2001) [reply tensorflow/python/keras/metrics.py:3346]: I'll do that 👍

File: tensorflow/python/keras/metrics.py [hunk 3]
```diff
@@ -3342,6 +3342,9 @@ def categorical_accuracy(y_true, y_pred):
   Returns:
     Categorical accuracy values.
   """
+  # assert if predicted and true labels Tensors have the same shape
+  assert tf.shape(y_true)==y_pred.shape,"Both `y_true` and `y_pred` Tensors must have same shapes and must be one-hot encoded. Use tf.one_hot() to one hot encode the inputs."
```
Comment (AdityaKane2001) [inline tensorflow/python/keras/metrics.py:3346]: Made requested changes.

**Review by AdityaKane2001 — COMMENTED** [review]
Made requested changes.

Edit:
Build error. Changed `tf.shape()` to `array_ops.shape_v2()`

**Review by AdityaKane2001 — COMMENTED** [review]
`tf is not defined `. Used `tensorflow.python.ops.array_ops.shape_v2()` for calculating dynamic shape of `y_true` Tensor

**Review by AdityaKane2001 — COMMENTED** [review]
Used `array_ops.shape_v2()`

**Comment by AdityaKane2001** [comment]
Following test is raising the `AssertionError`:
` test_saving_with_sequence_features`
Source code [here](https://github.com/tensorflow/tensorflow/blob/362f9fc585905592094db83e4f677a0fc7a65972/tensorflow/python/keras/feature_column/sequence_feature_column_test.py#L617).
@chenmoneygithub  
How can I resolve these errors?
Thanks.

**Comment by chenmoneygithub** [comment]
Hi, sorry for the late reply. I tried your code by myself, and it will break the graph mode tests. The code below works:
```python
check_ops.assert_equal_v2(
      array_ops.shape_v2(y_pred), array_ops.shape_v2(y_pred))
```

You can copy & paste the code, and I will approve it. Thanks for your contribution to Keras!

**Comment by AdityaKane2001** [comment]
@chenmoneygithub 
Did you mean 
```
check_ops.assert_equal_v2(
      array_ops.shape_v2(y_pred), array_ops.shape_v2(y_true))
```

**Comment by chenmoneygithub** [comment]
Yes, sorry for the typo.

**Comment by AdityaKane2001** [comment]
@chenmoneygithub 

Thanks a lot for the support!

**Review by chenmoneygithub — APPROVED** [review]
thanks for the PR!

**Comment by gbaned** [comment]
@AdityaKane2001 Can you please fix build failures ? Thanks!

**Comment by AdityaKane2001** [comment]
@gbaned
Can you please tell how can I change the necessary test cases? 
I can locate the files which are causing the errors, but I am not able to get to the actual test cases.

**Comment by AdityaKane2001** [comment]
@gbaned  @chenmoneygithub 
I believe this time around it's breaking for eager mode.
After my last commit, there were no tests done for Windows and Ubuntu. How do I proceed from here?
Thanks,
Aditya

File: tensorflow/python/keras/layers/local_test.py [hunk 4]
```diff
@@ -288,8 +288,10 @@ def test_locallyconnected_implementation(self, width, data_format):
 
       np.random.seed(1)
       tf_test_util.random_seed.set_seed(1)
-      targets = np.random.randint(0, num_classes, (num_samples,))
-
+      sparse_targets = np.random.randint(0, num_classes, (num_samples,)) 
+      targets =  np.zeros((sparse_targets.size, sparse_targets.max()+1))
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/local_test.py:292]: Nit: space surround operator => sparse_targets.max() + 1

File: tensorflow/python/keras/layers/local_test.py [hunk 5]
```diff
@@ -288,8 +288,10 @@ def test_locallyconnected_implementation(self, width, data_format):
...
+      targets[np.arange(sparse_targets.size),sparse_targets] = 1
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/local_test.py:293]: Nit: whitespace after comma => , sparse_targets

File: tensorflow/python/keras/layers/local_test.py [hunk 6]
```diff
@@ -288,8 +288,10 @@ def test_locallyconnected_implementation(self, width, data_format):
 
       np.random.seed(1)
       tf_test_util.random_seed.set_seed(1)
-      targets = np.random.randint(0, num_classes, (num_samples,))
-
+      sparse_targets = np.random.randint(0, num_classes, (num_samples,))
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/local_test.py:291]: Maybe add comments for unintuitive code blocks. These 3 lines create a matrix with each row a one-hot vector, but it's not that obvious.

Comment (AdityaKane2001) [reply tensorflow/python/keras/layers/local_test.py:291]: Added comment and made requested changes

File: tensorflow/python/keras/layers/local_test.py [hunk 7]
```diff
@@ -288,8 +288,24 @@ def test_locallyconnected_implementation(self, width, data_format):
...
+      # to one-hot encoded vectors
+    
+      # Create sparse targets eg. [0,1,2]
+      sparse_targets = np.random.randint(0, num_classes, (num_samples,)) 
+      
+      # Convert to one-hot encoding
+      targets =  np.zeros((sparse_targets.size, sparse_targets.max() + 1)) 
+      targets[np.arange(sparse_targets.size), sparse_targets] = 1 
+      
+      # Final targets:
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/local_test.py:302]: Please put this comment block after "# Convert to one-hot encoding", and to semantically separate code blocks, just use one empty line. Thanks!

Comment (AdityaKane2001) [reply tensorflow/python/keras/layers/local_test.py:302]: The `Final targets :` one?

Comment (AdityaKane2001) [reply tensorflow/python/keras/layers/local_test.py:302]: Made requested changes

File: tensorflow/python/keras/layers/local_test.py [hunk 8]
```diff
@@ -288,8 +288,23 @@ def test_locallyconnected_implementation(self, width, data_format):
 
       np.random.seed(1)
       tf_test_util.random_seed.set_seed(1)
-      targets = np.random.randint(0, num_classes, (num_samples,))
-
+      
+      # Following code generates sparse targets and converts them 
+      # to one-hot encoded vectors
+
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/local_test.py:294]: Please remove this blank line.

File: tensorflow/python/keras/layers/local_test.py [hunk 9]
```diff
@@ -288,8 +288,23 @@ def test_locallyconnected_implementation(self, width, data_format):
...
-
+      
+      # Following code generates sparse targets and converts them 
+      # to one-hot encoded vectors
+    
+      # Create sparse targets eg. [0,1,2]
+      sparse_targets = np.random.randint(0, num_classes, (num_samples,)) 
+      
+      # Convert to one-hot encoding
+
```
Comment (chenmoneygithub) [inline tensorflow/python/keras/layers/local_test.py:299]: Please remove this blank line too.

**Review by chenmoneygithub — APPROVED** [review]
In all looks good, thanks for the contribution!

File: tensorflow/python/keras/layers/local_test.py (hunk 9 above)
Comment (AdityaKane2001) [reply tensorflow/python/keras/layers/local_test.py:299]: Made requested changes
Thanks for the support!

Comment (AdityaKane2001) [reply tensorflow/python/keras/layers/local_test.py]: The tests didn't complete, perhaps due to the latest commit.

Comment (AdityaKane2001) [reply tensorflow/python/keras/layers/local_test.py]: The `tensorflow-butler` bot removed `ready to pull` label. I hope the PR will get merged 😅

**Comment by AdityaKane2001** [comment]
@gbaned @chenmoneygithub 
This PR was originally intended to assert the shapes in categorical accuracy function, as mentioned in the [issue](https://github.com/tensorflow/tensorflow/issues/46953). But, there are a few errors in `tensorflow/tensorflow/python/keras/layers/local_test.py` file, which cause 3 of the tests to fail. 
So I intend to do the following:
1. Use this PR to just change the `categorical_accuracy()` function. 
2. I'll create a fresh PR which takes care of the errors in the aforementioned file.

I'll proceed to do the same if it's good by you.

Thanks again for the support,
Aditya Kane

**Comment by chenmoneygithub** [comment]
Thanks Aditya!
